pip/pip-278.md

```diff
@@ -0,0 +1,158 @@
+# Background knowledge
+
+Apache Pulsar is a distributed messaging system that supports multiple messaging protocols and storage methods. 
+Among them, Pulsar Topic Compaction provides a key-based data retention mechanism that allows you only to keep the most recent message associated with that key to reduce storage space and improve system efficiency.
+
+Another Pulsar's internal use case, the Topic Compaction of new load balancer changed the strategy of compaction, it will only to keep the first value of the key. More detail please see: [PIP-215](https://github.com/apache/pulsar/issues/18099)
+
+More topic compaction details can be found in [Pulsar Topic Compaction](https://pulsar.apache.org/docs/en/concepts-topic-compaction/).
+
+# Motivation
+
+Currently, the implementation of Pulsar Topic Compaction is fixed and does not support custom strategy, which limits users from using more Compactor policies in their applications.
+
+For example, current topic compaction can work with pulsar format data in KoP, but it can't work with Kafka format data since the data written to the entry is in Kafka format. 
+The Pulsar compactor doesn't aware of Kafka format data. And it doesn't make sense to support Kafka format data handling in Pulsar. We need implement a pluggable compactor to support Kafka format data handling in KoP.
+
+Another long-term consideration is that we may need to support write the compacted data anywhere, S3, in columnar format, or even partitioning.
+
+So we need to make  the whole topic compaction service (include Write API & Read API) pluggable to support more customize compaction service implementation.
+
+# Goals
+
+## In Scope
+
+* Abstract topic compaction service interface and support topic compaction service pluggable.
+
+* Migrate the current implementation to a new interface implementation.
+
+* Makes existing tests compatible with new implementations.
+
+## Out of Scope
+
+* For CompactorMetrics, keep the current implementation and don't define related method in topic compaction service interface, in the future will use the `Otel` interface or other mertics API instead.
```

Comment:
```suggestion
* For CompactorMetrics, keep the current implementation and don't define related methods in the topic compaction service interface. In the future, it will use the `Otel` interface or other metrics API instead.
```